packages/block-editor/src/components/content-only-controls/index.js

```diff
@@ -0,0 +1,176 @@
+/**
+ * WordPress dependencies
+ */
+import { store as blocksStore } from '@wordpress/blocks';
+import {
+	__experimentalToolsPanel as ToolsPanel,
+	__experimentalToolsPanelItem as ToolsPanelItem,
+	__experimentalHStack as HStack,
+	TextControl,
+} from '@wordpress/components';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { __unstableStripHTML as stripHTML } from '@wordpress/dom';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../../store';
+import BlockIcon from '../block-icon';
+import useBlockDisplayTitle from '../block-title/use-block-display-title';
+import useBlockDisplayInformation from '../use-block-display-information';
+
+function RichTextControl( { label, value, setValue } ) {
+	return (
+		<TextControl
+			__nextHasNoMarginBottom
+			__next40pxDefaultSize
+			label={ label }
+			value={ value ? stripHTML( value ) : '' }
+			onChange={ setValue }
+			autoComplete="off"
+		/>
+	);
+}
+
+function getControlForAttribute( attribute ) {
+	if ( attribute.control.type === 'RichText' ) {
+		return RichTextControl;
+	}
+}
+
+function getDefaultValue( attribute ) {
+	if ( attribute.default ) {
+		return attribute.default;
+	}
+
+	return undefined;
+}
+
+function BlockAttributeToolsPanelItem( {
+	clientId,
+	attributeDefinition,
+	setValue,
+	value,
+} ) {
+	const Control = getControlForAttribute( attributeDefinition );
+
+	if ( ! Control ) {
+		return null;
+	}
+
+	const defaultValue = getDefaultValue( attributeDefinition );
+
+	return (
+		<ToolsPanelItem
+			panelId={ clientId }
+			label={ attributeDefinition.control.label }
+			hasValue={ () => value !== defaultValue }
+			onDeselect={ () => setValue( defaultValue ) }
+			isShownByDefault={ attributeDefinition.control.shownByDefault }
+		>
+			<Control
+				label={ attributeDefinition.control.label }
+				value={ value }
+				setValue={ setValue }
+			/>
+		</ToolsPanelItem>
+	);
+}
+
+function getContentAttributesWithControls( blockType ) {
+	return Object.keys( blockType?.attributes ?? {} )
+		.filter( ( attributeKey ) => {
+			const attribute = blockType.attributes[ attributeKey ];
+			return attribute?.role === 'content' && !! attribute.control;
+		} )
+		.map( ( attributeKey ) => ( {
+			key: attributeKey,
+			...blockType?.attributes[ attributeKey ],
+		} ) );
+}
+
+function getResetAllValue( attributes ) {
+	const resetValue = {};
+
+	attributes.forEach( ( attribute ) => {
+		resetValue[ attribute.key ] = getDefaultValue( attribute );
```

Comment:
I think the reset options in the tools panel would be better if they reset to the pattern's original content (similar to how synced patterns work) rather than the attribute's default.

I haven't explored it yet, but I think it'll be pretty difficult to do. 😬